hooks/migrate_csi_endpoint.sh

```diff
@@ -0,0 +1,317 @@
+  export old_driver_name="linstor.csi.linbit.com"
+  export new_driver_name="replicated.csi.storage.deckhouse.io"
+  export old_attacher="linstor-csi-linbit-com"
+  export new_attacher="replicated-csi-storage-deckhouse-io"
+  export LABEL_KEY="storage.deckhouse.io/need-kubelet-restart"
+  export LABEL_VALUE=""
+  export SECRET_NAME="csi-migration-finished"
+  export NAMESPACE="d8-sds-replicated-volume"
+  export timestamp="$(date +"%Y%m%d%H%M%S")"
+  export affected_pvs_hash=""
+  export WEBHOOK_NAME="d8-sds-replicated-volume-sc-validation"
+  echo "Migration csi shell hook started"
+
+  if kubectl get secret "$SECRET_NAME" -n "$NAMESPACE" > /dev/null 2>&1; then
+    echo "Secret ${NAMESPACE}/${SECRET_NAME} exists. Migration has been finished before"
+    values::set sdsReplicatedVolume.internal.csiMigrationHook.completed "true"
+    exit 0
+  fi
+
+  echo "Secret ${NAMESPACE}/${SECRET_NAME} does not exist. Starting csi migration"
+
+  sc_list=$(kubectl get sc -o json | jq -r ".items[] | select(.provisioner == \"$old_driver_name\") | .metadata.name")
+  pvc_pv_list=$(kubectl get pv -o json | jq -r --arg oldDriverName "$old_driver_name" '.items[] | select(.spec.csi.driver == $oldDriverName) | .spec.claimRef.namespace + "/" + .spec.claimRef.name + "/" + .metadata.name')
+  pvc_list=$(kubectl get pvc --all-namespaces -o json | jq -r --arg oldDriverName "$old_driver_name" '.items[] | select(.metadata.annotations["volume.kubernetes.io/storage-provisioner"] == $oldDriverName) | .metadata.namespace + "/" + .metadata.name')
+  
+  if [[ -z "$sc_list" && -z "$pvc_pv_list" && -z "$pvc_list" ]]; then
+    echo "No StorageClasses and PVCs/PVs to migrate. Migration not needed"
+    values::set sdsReplicatedVolume.internal.csiMigrationHook.completed "true"
+    kubectl -n ${NAMESPACE} create secret generic ${SECRET_NAME}
+    exit 0
+  fi
+  
+  delete_resource ${NAMESPACE} daemonset linstor-csi-node
+  scale_down_pods ${NAMESPACE} linstor-csi-controller
+  scale_down_pods ${NAMESPACE} linstor-affinity-controller
+  scale_down_pods ${NAMESPACE} linstor-scheduler
+  scale_down_pods ${NAMESPACE} sds-replicated-volume-controller
+  scale_down_pods ${NAMESPACE} linstor-scheduler-admission
+
+
+  export temp_dir=$(mktemp -d)
+  cd "$temp_dir"
+  echo "temporary dir: $temp_dir"
+
+  migrate_storage_classes
+  migrate_pvc_pv
+  migrate_volume_snapshot_classes
+  migrate_volume_snapshot_contents
+  
+  nodes_with_volumes=$(kubectl get volumeattachments -o=jsonpath='{range .items[*]}{@.metadata.name}{"\t"}{@.spec.attacher}{"\t"}{@.status.attached}{"\t"}{@.spec.nodeName}{"\n"}{end}' | grep "${old_driver_name}" | grep 'true' | awk '{print $4}' | sort | uniq)
+  echo nodes_with_volumes=$nodes_with_volumes
+
+  delete_old_volume_attachments
+
+
+  for node in $nodes_with_volumes; do
+    echo "Add label ${LABEL_KEY}=${LABEL_VALUE} to node $node"
+    kubectl label node $node ${LABEL_KEY}=${LABEL_VALUE} --overwrite
+  done
+
+  echo affected_pvs_hash="$affected_pvs_hash"
```

Comment:
isn't obvious where was var set.

Reply:
Move export AFFECTED_PVS_HASH="" to a position before the functions where it is set.